Added CLI to run pysr.install() to install Julia dependencies

w2ll2am opened the pull request:
This PR implements at CLI for the pysr project using the [click](https://click.palletsprojects.com/en/8.1.x/) package and responds to [issue-229](https://github.com/MilesCranmer/PySR/issues/229). The CLI is intended to be used as `python -m pysr install`. Included in this PR is the following:

1. Setup of the `install` command and it's associated options
2. Including click` in dependencies for the package
3. Updating install instructions to include `python -m pysr install` in place of `python -c "import pysr; pysr.install()"`

Please do leave comments below as verbose as possible, as I'm new to this community.

MilesCranmer commented on pysr/_cli/main.py:
```diff
@@ -0,0 +1,16 @@
+    ctx = context
+
+
+@cli.command("install", help="Install required Julia dependencies")
```
```suggestion
@cli.command("install", help="Install Julia dependencies for PySR")
```

MilesCranmer commented on pysr/_cli/main.py:
```diff
@@ -0,0 +1,16 @@
+
+
+@cli.command("install", help="Install required Julia dependencies")
+@click.option("--julia_project", default=None)
```
Should specify the type is a string

MilesCranmer replied:
```suggestion
@click.option("--julia-project", default=None, type=str)
```

MilesCranmer commented on pysr/_cli/main.py:
```diff
@@ -0,0 +1,16 @@
+@cli.command("install", help="Install required Julia dependencies")
+@click.option("--julia_project", default=None)
+@click.option("--quiet", default=False)
+@click.option("--precompile", default=None)
```
I think this should be a feature flag: https://click.palletsprojects.com/en/8.1.x/options/#feature-switches

where `--precompile` => `True`, `--no-precompile` => `False`, and otherwise => `None`.

w2ll2am replied:
Thank you I didn't realise there were three different inputs for precompile.

MilesCranmer commented on pysr/_cli/main.py:
```diff
@@ -0,0 +1,16 @@
+
+@cli.command("install", help="Install required Julia dependencies")
+@click.option("--julia_project", default=None)
+@click.option("--quiet", default=False)
```
```suggestion
@click.option("--quiet", is_flag=True, default=False)
```

MilesCranmer reviewed (commented):
Looks great! Added some comments. One other suggestion is regarding the file structure. Is `_cli` rather than `cli` a choice because of naming issue or is it the standard?

MilesCranmer commented:
Btw, I think your commits are associated with an email address associated to @WilliamBC-SL rather than @w2ll2am. Just a heads up in case you want the PR credited to the latter account; either one is fine with me. You can change this with https://stackoverflow.com/questions/750172/how-do-i-change-the-author-and-committer-name-email-for-multiple-commits

w2ll2am commented:
> Looks great! Added some comments. One other suggestion is regarding the file structure. Is `_cli` rather than `cli` a choice because of naming issue or is it the standard?

I was basing the structure off pip's CLI which stores its `main.py` for the CLI in the following directory: [`pip/_internal/cli/main.py`](https://github.com/pypa/pip/blob/main/src/pip/_internal/cli/main.py). My understanding is that the underscore prefix can be used for methods or files that should be for internal use of the module [detailed in this article](https://dbader.org/blog/meaning-of-underscores-in-python#:~:text=It's%20a%20hint%20to%20the,is%20intended%20for%20internal%20use.). I'm happy to choose the structure that is the best fit for the project.

MilesCranmer commented on requirements.txt:
```diff
@@ -3,3 +3,5 @@ pandas
 numpy
 scikit_learn >= 1.0.0
 julia >= 0.6.0
+click>=8.0.4
```
```suggestion
click>=7.0.0
```

Looks like the old version has the same API so we could lower this (in case another package in the same env requires an old version).

MilesCranmer commented on requirements.txt:
```diff
@@ -3,3 +3,5 @@ pandas
 scikit_learn >= 1.0.0
 julia >= 0.6.0
+click>=8.0.4
+setuptools>=66.0.0
```
```suggestion
setuptools>=50.0.0
```
(Not sure if the constraint is needed)

MilesCranmer commented:
Looks almost ready! One more request - update the usage string. Right now it shows this:

```
> python -m pysr --help
Usage: cli install [OPTIONS]

  Install Julia dependencies for PySR

Options:
  -p, --project TEXT
  -q, --quiet
  --precompile
  --no-precompile
  --help              Show this message and exit.
```

I think this should be, for example (the red texts are deletions, green text additions):


```diff
  python -m pysr --help
- Usage: cli install [OPTIONS] 
+ Usage: pysr install [OPTIONS]
    Install Julia dependencies for PySR
  
  Options:
-    -p, --project TEXT
+    -p, --project PROJECT_DIRECTORY      Install in a specific Julia project (e.g., a local copy of SymbolicRegression.jl).
-   -q, --quiet
+   -q, --quiet          Disable logging.
-    --precompile
+    --precompile        Force precompilation of Julia libraries.
-    --no-precompile
+    --no-precompile     Disable precompilation.
    --help              Show this message and exit.
```

WilliamBC-SL commented:
> Looks almost ready! One more request - update the usage string. Right now it shows this:
> 
> ```
> > python -m pysr --help
> Usage: cli install [OPTIONS]
> 
>   Install Julia dependencies for PySR
> 
> Options:
>   -p, --project TEXT
>   -q, --quiet
>   --precompile
>   --no-precompile
>   --help              Show this message and exit.
> ```
> 
> I think this should be, for example (the red texts are deletions, green text additions):
> 
> ```diff
>   python -m pysr --help
> - Usage: cli install [OPTIONS] 
> + Usage: pysr install [OPTIONS]
>     Install Julia dependencies for PySR
>   
>   Options:
> -    -p, --project TEXT
> +    -p, --project PROJECT_DIRECTORY      Install in a specific Julia project (e.g., a local copy of SymbolicRegression.jl).
> -    -q, --quiet
> +    -q, --quiet          Disable logging.
> -    --precompile
> +    --precompile        Force precompilation of Julia libraries.
> -    --no-precompile
> +    --no-precompile     Disable precompilation.
>     --help              Show this message and exit.
> ```

Of course! Implemented now.

MilesCranmer commented:
Sorry for the delay; I am trying to think about how we could have unit-tests for this besides the install step itself. But I think it is ready to be merged, nice job!

w2ll2am commented:
In this PR I changed all of the `pysr.install()` lines to use the CLI instead, so all of the deepsource tests rely on `python -m pysr install` to work and pass. So the functionality is tested through that, and I've just added a PR to test the `--help` messages.

w2ll2am commented:
I'm happy for the code to be merged when you are!

MilesCranmer commented:
Sorry for the long delay! Has been a crazy couple of weeks. 

I think the unittests aren't actually running yet, because it needs to be added to the testing CLI: https://github.com/MilesCranmer/PySR/blob/master/pysr/test/__init__.py and https://github.com/MilesCranmer/PySR/blob/master/pysr/test/__main__.py

MilesCranmer commented:
Could probably add a "cli" option here: https://github.com/MilesCranmer/PySR/blob/60c32792958e9b357d12716b8e614cdc6949a50b/pysr/test/__main__.py#L28

w2ll2am commented:
I'm not sure why these tests are failing? 
![image](https://github.com/MilesCranmer/PySR/assets/16038228/ed231d4d-5f4c-4be7-a992-c678970749c1)

MilesCranmer commented:
Okay I think I figured out why it’s still not testing. I think it just needs to be added to the GitHub actions; like: `python -m pysr cli` to run the CLI tests.

MilesCranmer commented:
I pushed directly to the branch; hope that's okay with you. Eager to get this working.

w2ll2am commented:
Thank you I'm very happy with you making changes! I'm all happy with the PR.

MilesCranmer commented:
Thanks again for the amazing PR!! I have added you to the contributors list 💯
